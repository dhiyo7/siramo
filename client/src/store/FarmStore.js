import { observable, computed } from 'mobx'
import { db, User } from './firebase'
import { Alert } from 'react-native'


class FarmStore {
  @observable farmData = {
    historyHumidity: [],
    historyWaterRatio: [],
    historyTemperature: [],
    historyWaterLevel: [],
    historyLoading: false
  }

  @observable navigation = {}

  getHistory = (key) => {
    this.farmData.historyLoading = true
    db.ref('/history/17UFak7JqufG1RXUeVW30jwdfrQ2').on('value', (snapshot) => {

      snapshot.forEach(snap => {
        this.farmData.historyHumidity.push({
          x: snap.val().last_updated,
          y: parseFloat(snap.val().humidity)
        })
        this.farmData.historyWaterRatio.push({
          x: snap.val().last_updated,
          y: Math.round(parseFloat(snap.val().water_ratio))
        })
        this.farmData.historyTemperature.push({
          x: snap.val().last_updated,
          y: parseFloat(snap.val().temperature)
        })
        this.farmData.historyWaterLevel.push({
          x: snap.val().last_updated,
          y: parseFloat(snap.val().water_level)
        })
      })
      this.farmData.historyLoading = false
    })
  }

  setSchedule = (cronFormat, max, min) => {
    let updatedValue = {
      name: cronFormat,
      maxWaterRatio: max,
      minWaterRatio: min
    }
    console.log('Firebase Updated', updatedValue)
    db.ref(`/users`).update(updatedValue)
      .then(response => {
        console.log(response)
      })
      .catch(err => {
        console.log(err)
      })
  }

}

export default new FarmStore()