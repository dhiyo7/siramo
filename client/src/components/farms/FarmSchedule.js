import React, { Component } from 'react'
import {
  View, Text, TouchableOpacity, 
  TextInput, StyleSheet, Picker, Switch
} from 'react-native'
import { Ionicons } from '@expo/vector-icons'
import { inject, observer } from 'mobx-react'
import FarmStore from '../../store/FarmStore'

@observer class FarmSchedule extends Component {
  constructor() {
    super()
    this.state = {
      automaticMode: false,
      minutes: [
        '00','01','02','03','04','05','06','07','08', '09',
        '10','11','12','13','14','15','16','17','18', '19',
        '20','21','22','23','24','25','26','27','28', '29',
        '30','31','32','33','34','35','36','37','38', '39',
        '40','41','42','43','44','45','46','47','48', '49',
        '50','51','52','53','54','55','56','57','58', '59'
      ],
      hours: [
        '00','01','02','03','04','05','06','07','08', '09',
        '10','11','12','13','14','15','16','17','18', '19',
        '20','21','22','23'
      ],
      waterRatio: [
        1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
        21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 33, 35, 36, 37, 38, 39, 40,
        41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60,
        61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
        81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 99, 98, 99, 100
      ],
      hoursPick: '00',
      minutesPick: '00',
      maxWaterRatio: 0,
      minWaterRatio: 0
    }
  }
  
  componentDidMount() {
    if (FarmStore.FarmDetail.cronSchedule) {
      let splitCron = FarmStore.FarmDetail.cronSchedule.split(' ')
      let minutes = splitCron[1].match(/[0-9]/g).join('').toString()
      let hours = splitCron[2].match(/[0-9]/g).join('').toString()
      
      this.setState({
        minWaterRatio: FarmStore.FarmDetail.minWaterRatio,
        maxWaterRatio: FarmStore.FarmDetail.maxWaterRatio,
        minutesPick: (minutes < 10)? `0${minutes}`:`${minutes}`,
        hoursPick: (hours < 10)? `0${hours}`:`${hours}`
      })
    }
  }

  startSchedule = () => {
    let cronFormat = "3 59 */23 * * *" // setiap jam 00:00 tengah malam
    let { 
      hoursPick, minutesPick, maxWaterRatio, minWaterRatio
    } = this.state
    if (hoursPick === '00' && minutesPick !== '00') {
      cronFormat = `3 */${parseInt(minutesPick)} * * * *`
    } else if (hoursPick !== '00') {
      cronFormat = `3 ${parseInt(minutesPick)} */${parseInt(hoursPick)} * * *`
    }
    console.log('Jam', this.state.hoursPick)
    console.log('Menit', this.state.minutesPick)
    console.log('Max', this.state.maxWaterRatio)
    console.log('min', this.state.minWaterRatio)

    FarmStore.setSchedule(cronFormat, maxWaterRatio, minWaterRatio)
  }
  
  changeMode () {
    if(this.state.automaticMode) {
      this.setState({
        automaticMode: false
      })
      FarmStore.changeMode(false)
    } else {
      this.setState({
        automaticMode: true
      })
      FarmStore.changeMode(true)
    }
  }
  render() {
    const { 
      hours, minutes, hoursPick, minutesPick,
      maxWaterRatio, minWaterRatio, waterRatio
    } = this.state
    const hoursFormat = hours.map((hour, index) => (
      <Picker.Item label={hour} value={hour} key={index}/>
    ))
    const minutesFormat = minutes.map((minute, index) => (
      <Picker.Item label={minute} value={minute} key={index}/>
    ))
    const wtrRatio = waterRatio.map((wr, index) => (
      <Picker.Item label={`${wr}`} value={wr} key={index}/>
    ))

    const automaticModeOn = 
    <View style={styles.container}>
      <Text style={styles.txtTitle}>Watering Schedule</Text>
        <View style={styles.inputBox}>
        <Text><Ionicons name='md-clock' size={20} />  Hours</Text>
        <Picker
          style={styles.timeInput}
          selectedValue={hoursPick}
          onValueChange={(itemValue, itemIndex) => this.setState({hoursPick: itemValue})}
        >
          {hoursFormat}
        </Picker>
        <Text>Minutes</Text>
        <Picker
          style={styles.timeInput}
          selectedValue={minutesPick}
          onValueChange={(itemValue, itemIndex) => this.setState({minutesPick: itemValue})}
        >
          {minutesFormat}
        </Picker>
      </View>

      <View style={styles.inputBox}>
        <Text><Ionicons name='md-rainy' size={20} /> Maximum Water Ratio</Text>
        <Picker
          style={styles.timeInput}
          selectedValue={maxWaterRatio}
          onValueChange={(value, index) => this.setState({maxWaterRatio: value})}
        >
          {wtrRatio}
        </Picker>
      </View>

      <View style={styles.inputBox}>
        <Text><Ionicons name='md-cloud-outline' size={20} /> Minimum Water Ratio</Text>
        <Picker
          style={styles.timeInput}
          selectedValue={minWaterRatio}
          onValueChange={(value, index) => this.setState({minWaterRatio: value})}
        >
          {wtrRatio}
        </Picker>
      </View>

      <TouchableOpacity
          style={styles.buttonContainer}
          onPress={this.startSchedule}
        >
          <Text style={styles.buttonText}>Save Change</Text>
        </TouchableOpacity>
    </View>

    return (
      <View style={styles.mainContainer}>
        <View style={styles.switchContainer}>
          <View style={{width: 280}}>
            <Text style={{fontSize: 20}}> Automatic Mode </Text>
          </View>
          <View>
            <Switch 
                  onValueChange = {() => this.changeMode()}
                  value = {this.state.automaticMode}
                />
          </View>
        </View>
      <View>
          {
            this.state.automaticMode?
            automaticModeOn: <Text style={styles.container}> </Text>
          }
      </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  mainContainer: {
    backgroundColor: '#F1F8E9',
    paddingLeft: 12,
    paddingRight: 12,
    paddingTop: 12,
    height: '100%'
  },
  container: {
    backgroundColor: '#F1F8E9',
    width: '100%',
    marginRight: 8,
    marginLeft: 8,
    marginTop: 10,
    alignItems: 'center'
  },
  txtTitle: {
    fontSize: 30,
    fontWeight: 'bold',
  },
  inputBox: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 4,
    marginTop: 4,
    width: 400,
  },
  iconBox: {
    width: 25,
    height: 25,
    justifyContent: 'center',
    alignItems: 'center',
  },
  timeInput: {
    height: 40,
    width: 40,
    backgroundColor: 'rgba(225,225,225,0.2)',
    padding: 6,
    marginRight: 4,
    marginLeft: 4,
    marginTop: 2,
    marginBottom: 2,
    color: '#8D6E63',
    borderRadius: 4,
  },
  buttonContainer: {
    backgroundColor: '#33691E',
    paddingVertical: 15,
    marginBottom: 8,
    borderRadius: 4,
    width: 320,
    alignItems: 'center',
    justifyContent: 'center'
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
    fontWeight: '700'
  },
  switchContainer: {
    width: '100%',
    backgroundColor: '#F5F5F5',
    flexWrap: 'wrap',
    flexDirection: 'row',
    display: 'flex',
    borderRadius: 6,
    borderBottomWidth: 1,
    borderBottomColor: '#795548',
    borderTopWidth: 1,
    borderTopColor: '#795548', 
    borderRightWidth: 1,
    borderRightColor: '#795548',
    borderLeftWidth: 1,
    borderLeftColor: '#795548',
  },
})

export default FarmSchedule