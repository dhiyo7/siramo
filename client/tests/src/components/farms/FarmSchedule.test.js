import React from 'react'
import {
  View, Text, TouchableOpacity, 
  TextInput, StyleSheet, Picker
} from 'react-native'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { Ionicons } from '@expo/vector-icons'

import FarmStore from '../../../../src/store/FarmStore'
import FarmSchedule from '../../../../src/components/farms/FarmSchedule'
import userStore from '../../../../src/store/UserStore';

Enzyme.configure({ adapter: new Adapter()})

let wrapper

beforeEach(() => {
  wrapper = shallow(<FarmSchedule />)
})

describe('<FarmSchedule /> snapshot Testing', () => {
  it('snapshot works!', () => {
    const tree = renderer.create(<FarmSchedule />)
    expect(tree).toMatchSnapshot()
  })
})

describe('<FarmSchedule />', () => {
  it('should render <FarmDetail />', () => {
    expect(wrapper).toBeDefined()
  })
})

describe('<FarmSchedule /> state', () => {
  it('should have state value', () => {
    expect(wrapper.state('minutes').length).toBe(60)
    expect(typeof wrapper.state('minutes')[0]).toBe('string')
    expect(wrapper.state('hours').length).toBe(24)
    expect(typeof wrapper.state('hours')[0]).toBe('string')
    expect(wrapper.state('waterRatio').length).toBe(100)
    expect(typeof wrapper.state('waterRatio')[0]).toBe('number')
    expect(wrapper.state('hoursPick')).toBe('00')
    expect(wrapper.state('minutesPick')).toBe('00')
    expect(wrapper.state('maxWaterRatio')).toBe(0)
    expect(wrapper.state('minWaterRatio')).toBe(0)
  })
})

describe('<FarmSchedule /> child is rendered', () => {
  it('should render <FarmSchedule /> child Element', () => {
    expect(wrapper.find(<View />)).toBeDefined()
    expect(wrapper.find(<Text />)).toBeDefined()
    expect(wrapper.find('Text').length).toBe(6)
    expect(wrapper.find(<Picker />)).toBeDefined()
    expect(wrapper.find('Picker').length).toBe(4)
    expect(wrapper.find(<Picker.Item />)).toBeDefined()
    expect(wrapper.find(<TouchableOpacity />)).toBeDefined()
    expect(wrapper.find('TouchableOpacity').length).toBe(1)
  })
})

describe('<FarmSchedule /> <Text /> testing', () => {
  it('should have expected childern', () => {
    expect(wrapper.find('Text').get(0).props.children).toEqual('Watering Schedule')
    expect(wrapper.find('Text').get(1).props.children[0]).toEqual(<Ionicons name='md-clock' size={20} />)
    expect(wrapper.find('Text').get(1).props.children[1]).toEqual('  Hours')
    expect(wrapper.find('Text').get(2).props.children).toEqual('Minutes')
    expect(wrapper.find('Text').get(3).props.children[0]).toEqual(<Ionicons name='md-rainy' size={20} />)
    expect(wrapper.find('Text').get(3).props.children[1]).toEqual(' Maximum Water Ratio')
    expect(wrapper.find('Text').get(4).props.children[0]).toEqual(<Ionicons name='md-cloud-outline' size={20} />)
    expect(wrapper.find('Text').get(4).props.children[1]).toEqual(' Minimum Water Ratio')
    expect(wrapper.find('Text').get(5).props.children).toEqual('Save Change')
  })
})

describe('<FarmSchedule /> <Picker /> testing', () => {
  it('should have Picker.Item according to state', () => {
    expect(wrapper.find('Picker').get(0).props.children.length).toEqual(24)
    expect(wrapper.find('Picker').get(1).props.children.length).toEqual(60)
    expect(wrapper.find('Picker').get(2).props.children.length).toEqual(100)
    expect(wrapper.find('Picker').get(3).props.children.length).toEqual(100)
  })
})

describe('<FarmSchedule /> state Test on Change', () => {
  it('State value should have changed, when pickers changed', () => {
    wrapper.find('Picker').at(0).simulate('ValueChange', '04')
    expect(wrapper.state('hoursPick')).toEqual('04')
    wrapper.find('Picker').at(1).simulate('ValueChange', '23')
    expect(wrapper.state('minutesPick')).toEqual('23')
    wrapper.find('Picker').at(2).simulate('ValueChange', 90)
    expect(wrapper.state('maxWaterRatio')).toEqual(90)
    wrapper.find('Picker').at(3).simulate('ValueChange', 30)
    expect(wrapper.state('minWaterRatio')).toEqual(30)
  })
})

describe('<FarmSchedule /> componentDidMount Test', () => {

  it('state value should change according to FarmStore.FarmDetail data', () => {
    FarmStore.FarmDetail.cronSchedule = "* 2 12 * * *"
    FarmStore.FarmDetail.minWaterRatio = 30
    FarmStore.FarmDetail.maxWaterRatio = 80
    let tree = shallow(<FarmSchedule />)
    // let cronSchedule = "12"
    // console.log('------------------------------')
    // console.log(cronSchedule.match(/[0-9]/g).join(''))
    expect(tree.state('hoursPick')).toEqual('12') // "* 2 22 * * *" => 2,2 ?
    expect(tree.state('minutesPick')).toEqual('02')
    expect(tree.state('maxWaterRatio')).toEqual(80)
    expect(tree.state('minWaterRatio')).toEqual(30)
  })
})

describe('<FarmSchedule /> Button Save Change', () => {

  it('should trigger startSchedule in FarmStore', async() => {
    FarmStore.clearAll()
    userStore.userData.uid = '17UFak7JqufG1RXUeVW30jwdfrQ2'
    let tree = shallow(<FarmSchedule />)
    // let startSchedule = jest.fn()
    tree.find('Picker').at(0).simulate('ValueChange', '08')
    tree.find('Picker').at(1).simulate('ValueChange', '22')
    tree.find('Picker').at(2).simulate('ValueChange', 82)
    tree.find('Picker').at(3).simulate('ValueChange', 31)
    tree.find('TouchableOpacity').get(0).props.onPress()
    await FarmStore.getFarmData()
    expect(FarmStore.FarmDetail.cronSchedule).toEqual('* 22 */8 * * *')
    expect(FarmStore.FarmDetail.minWaterRatio).toEqual(31)
    expect(FarmStore.FarmDetail.maxWaterRatio).toEqual(82)
  })
})