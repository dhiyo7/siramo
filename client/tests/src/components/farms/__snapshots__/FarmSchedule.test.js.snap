// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FarmSchedule /> snapshot Testing snapshot works! 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F1F8E9",
      "height": "100%",
      "paddingTop": 12,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#C8E6C9",
        "borderBottomColor": "#00C853",
        "borderBottomWidth": 1,
        "borderLeftColor": "#00C853",
        "borderLeftWidth": 1,
        "borderRadius": 2,
        "borderRightColor": "#00C853",
        "borderRightWidth": 1,
        "borderTopColor": "#00C853",
        "borderTopWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "width": 300,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
         Automatic Mode 
      </Text>
    </View>
    <View>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            undefined,
          ]
        }
        value={false}
      />
    </View>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F8E9",
          "marginLeft": 8,
          "marginRight": 8,
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
       
    </Text>
  </View>
</View>
`;
